### ISSUE TITLE
Refactor `grad_accum` to `microbatch_size_per_device`

### ISSUE THREAD

**Issue by abhi-mosaic** [mosaicml/composer#845]
## 🚀 Feature Request
Change the Trainer to use `microbatch_size_per_device` instead of `grad_accum`, for better UX as `train_batch_size` and `n_devices` changes, and better hardware utilization for certain algorithms.

## Motivation
When we take a fixed workload with fixed `train_batch_size`, and change the number of devices we would like to run on (e.g. 1xA100 -> 8xA100 or 8xA100 -> 1xA100), we have to adjust the `grad_accum` in order to maintain good efficiency, or avoid OOMs, respectively.

Whenever we change the value of `train_batch_size`, as we would in a batch size sweep, we again have to change `grad_accum` for the same reasons.

An alternative to this would be for the fixed workload to specify `microbatch_size_per_device`, and then this field would never have to be changed as the batch size or number of devices changes, because the Trainer would compute `grad_accum = train_batch_size // microbatch_size_per_device // n_devices` internally wherever need be. (Right now we basically do the inverse.)

Another benefit of this is for algorithms such as `selective_backprop` that dynamically change the number of samples in the batch. Right now, if I set `grad_accum: 10` for vanilla training and turn on `selective_backprop` with `drop_ratio: 0.5`, then I am still splitting the pruned minibatch into 10 microbatches when I could be splitting it into 5. If the Trainer used `microbatch_size_per_device`, and passed only that value to the `split_batch` function, we could be smarter and split into only 5 chunks, and train at full utilization. @growlix is looking into accelerating NLP workloads with `selective_backprop`, where this could make a big difference.

## Additional context
Just wanted to mention that I am aware of the dynamic `grad_accum` PR #485 , but I don't think it fully solves the issue for situations like `selective_backprop` because we would either have to detect lowered memory usage or try to increase microbatch size periodically. I think a combination of the two ideas would work best.

With both features together, we could:
* set `microbatch_size_per_device: auto`, 
* which would be auto-calculated on the first step of training, 
* resulting in some state variable like `trainer.state.microbatch_size_per_device=8` 
* and then on step K, when `selective_backprop` is turned on, nothing has to be profiled or detected, instead
* our `split_batch(batch, microbatch_size_per_device)` function just receives a smaller batch than usual, measures its size, and splits it into fewer microbatches, each of size `microbatch_size_per_device`.

**Comment by ravi-mosaicml** [comment]
This makes sense. For the trainer, we pass in the dataloader (which yields batches of size `optimization_batch_size / num_devices`). We would need to modify the split batch function signature to take `num_samples` instead of `n_splits` as well.

For the hparams, I could see us also removing the `eval_batch_size` parameter, since gradient accumulation does not apply in evaluation, and instead the dataloader should yield batches of size `microbatch_size_per_device`.

**Comment by ravi-mosaicml** [comment]
Going to close this issue and move it to a design doc internally. See https://www.notion.so/Rename-grad_accum-to-microbatch_size_per_device-e44d249e825347cf97ae7e3424e055b2.
